Epic 8222 : WNPRC Purchasing - Receiver's page & data entry page update

labkey-bpatel opened the pull request:
#### Rationale
* Features for the Receiving clerk to be able to search line items by the vendor name, item description, purchase order number, Invoice Number, or confirmation number. Also, allow to update 'quantity received' for a particular line item and upload attachments such as packing slips.
Spec [here](https://docs.google.com/document/d/17S6R7jDa3E2JlN4w3Rz5ADLrimem3RcmgomFYB-2huI/edit#).

#### Related Pull Requests
* https://github.com/LabKey/ehrModules/pull/162

#### Changes
* Add Receiver's webpart
* Add action to render Receiver webpart
* Add Purchasing Receiver's link on main wnprcUnits page
* Remove Issues link from wnprcUnits and add to Purchase Admin page
* Add quantityReceived to the data entry page
* Conditionalize the data entry form based on user perms
* Handle qcstate upon quantity received

labkey-bpatel committed:
Add Receiver's webpart. Add action to render Receiver webpart. Create receiver's query. Add links to Request data entry.

labkey-bpatel committed:
Add link to Purchasing Receiver. Remove Issues link (now added to Purchase Admin page).

labkey-bpatel committed:
Add quantityReceived to the data entry form. Disable or don't show parts of data entry form based on user perm. Add page title. Styling and metadata updates.

labkey-bpatel committed:
Server side error for negative quantity & quantityReceived. Keep disabled input fields on one side so that disable icon (circle with slash across it) doesn't overflow on the enabled input.

labkey-bpatel committed:
Add return URLs. Add attachments to purchase admin queries.

labkey-bpatel committed:
Handle qcState. Receiver related changes. Layout changes depending on the user. Test setup update.

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -49,13 +51,32 @@ export const App: FC = memo(() => {
+    );
+    const [hasPurchasingInsertPermission, setHasPurchasingInsertPermission] = useState<boolean>(
+        getServerContext().user.canInsert
+    );
```
I don't think you need to set these into state.  Looks like they're only getting set on initialization anyway.  Suggested code here will also ensure if user permission changes while they are on this page, that will be reflected on the next render instead of only on initialization.
`const { isAdmin, canUpdate, canInsert } = getServerContext().user;`
or if you want to rename for clarity
`const { isAdmin: hasPurchasingAdminPermission, canUpdate: hasPurchasingUpdatePermission, canInsert: hasPurchasingInsertPermission } = getServerContext().user;`

labkey-bpatel replied:
good suggestion, done

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -49,13 +51,32 @@ export const App: FC = memo(() => {
+    );
 
     // equivalent to componentDidMount and componentDidUpdate (if with dependencies, then equivalent to componentDidUpdate)
     useEffect(() => {
```
Not critical but these promises in useEffect could be refactored to use async/await which is the more modern approach.

`useEffect(() => {
        // is fired on component mount
        const reqRowId = ActionURL.getParameter('requestRowId');
        setRequestId(reqRowId);

        (async () => {
            //get QCStates
            try {
                const qc = await getData('core', 'qcState', 'RowId, Label');
                const states = qc.map(val => {
                        if (val)
                        {
                            return QCStateModel.create({
                                rowId: val.RowId,
                                label: val.Label
                            });
                        }
                    });
                setQCStates(states);

                if (reqRowId)
                {
                    // get ehr_purchasing.purchasingRequests data
                    const filter = [Filter.create('rowId', reqRowId)];
                    const requests = await getData('ehr_purchasing', 'purchasingRequests', '*', undefined, filter);
                    if (requests && requests.length > 0)
                    {
                        setRequestOrderModel(
                            RequestOrderModel.create({
                                rowId: requests[0].rowId,
                                account: requests[0].otherAcctAndInves ? 'Other' : requests[0].account,
                                otherAcctAndInves: requests[0].otherAcctAndInves || undefined,
                                vendorId: requests[0].vendorId,
                                justification: requests[0].justification,
                                shippingInfoId: requests[0].shippingInfoId,
                                shippingAttentionTo: requests[0].shippingAttentionTo,
                                comments: requests[0].comments,
                                otherAcctAndInvesWarning: requests[0].otherAcctAndInves
                                    ? "Warning: Please add 'Account & Principal Investigator' value '" +
                                    requests[0].otherAcctAndInves +
                                    "' into the system."
                                    : undefined,
                            })
                        );

                       ...

            } catch (error) {
                console.error(error);
            }
        })();
    }, []);    
`

labkey-bpatel replied:
done

labkey-bpatel committed:
Title & layout changes. Move 'controlled substance' next to Item Descr.

labkey-martyp commented on WNPRC_Purchasing/src/client/actions.ts:
```diff
@@ -38,9 +38,43 @@ export function getData(
+    }
+
+    // otherwise, identify if all the line items are received
+    let receivedCount = 0;
+    lineItems.forEach((lineItem: LineItemModel)=> {
+        if (lineItem.quantityReceived >= lineItem.quantity) {
+            ++receivedCount;
+        }
+    });
```
You can use reduce for this.

`const receivedCount = lineItems.reduce((count, lineItem) => {
        if (lineItem.quantityReceived >= lineItem.quantity) {
            return count++;
        }
    }, 0);`

labkey-bpatel replied:
this was not working as expected, i tried couple other variants as well. Left it as is

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemRow.tsx:
```diff
@@ -47,52 +52,78 @@ export const LineItemRow: FC<LineItemProps> = memo(props => {
+                            isReadOnly={!isAdmin && canUpdate}
+                        />
+                    </Col>
+                    <Col xs={1}>
```
xs={1} is pretty small for an input.  Did you verify it holds up on smaller width screens?  Same with the inputs below.

labkey-bpatel replied:
It is as expected

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemRow.tsx:
```diff
@@ -47,52 +52,78 @@ export const LineItemRow: FC<LineItemProps> = memo(props => {
+                        </Col>
+                    }
+
+                    <Col xs={2}/>
+                    <Col xs={1}/>
```
I'm not sure exactly what the layout of the page is, but could this space be given to any other inputs?

labkey-bpatel replied:
1 is sufficient, otherwise all the inputs are too spaced out

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemsPanel.tsx:
```diff
@@ -50,67 +53,94 @@ export const LineItemsPanel: FC<Props> = memo(props => {
-                    <div className="panel-title">
-                        {(!hasRequestId && 'Specify Items') || (hasRequestId && 'Requested Items')}
+        <>
+            {
```
{ is only necessary when you're mixing JS into JSX (TS into TSX).  You shouldn't need this here.

labkey-bpatel replied:
removed

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemsPanel.tsx:
```diff
@@ -50,67 +53,94 @@ export const LineItemsPanel: FC<Props> = memo(props => {
-            {errorMsg && <div className="alert alert-danger">{errorMsg}</div>}
-        </Panel>
+                    <div>
+                        {
```
Can be removed

labkey-bpatel replied:
done

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemsPanel.tsx:
```diff
@@ -50,67 +53,94 @@ export const LineItemsPanel: FC<Props> = memo(props => {
+                                    <Col xs={1}>Quantity received</Col>
+                                }
+                            </Row>
+                        }
```
Can be removed

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemsPanel.tsx:
```diff
@@ -50,67 +53,94 @@ export const LineItemsPanel: FC<Props> = memo(props => {
+                        })}
+                    </div>
+                    <div>
+                        {
```
Can be removed

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemsPanel.tsx:
```diff
@@ -50,67 +53,94 @@ export const LineItemsPanel: FC<Props> = memo(props => {
+                                <>
+                                    <div>
+                                        <Row>
+                                            <Col xs={4}></Col>
+                                            <Col xs={1}></Col>
+                                            <Col xs={1}></Col>
```
This is kind of strange.  Can you not just do `<Col xs={6} />`?  There are other places doing this as well that at least should be collapsed to empty tag.

labkey-bpatel replied:
was trying to match the header cols, but yeah can use 6, now updated

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemsPanel.tsx:
```diff
@@ -50,67 +53,94 @@ export const LineItemsPanel: FC<Props> = memo(props => {
+                                                Total ($):
+                                            </Col>
+                                            <Col className="calc-total-val" xs={1}>
+                                                {formatCurrency(
+                                                    lineItems.reduce(function (accumulatedVal, lineItem) {
+                                                        return lineItem.unitCost && lineItem.quantity
+                                                            ? accumulatedVal + lineItem.unitCost * lineItem.quantity
+                                                            : accumulatedVal;
+                                                    }, 0)
```
I would pull this out of the JSX and make it a useMemo.  It's a lot of JS in JSX (which is a messy language to begin with), also it's getting recalculated each render.

labkey-bpatel replied:
done

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemsPanel.tsx:
```diff
@@ -50,67 +53,94 @@ export const LineItemsPanel: FC<Props> = memo(props => {
+                                        </Row>
+                                    </div>
+                                    <div className="add-item">
+                                        <span id="add-line-item-row" title="Add item" className="add-item-icon" onClick={onClickAddRow}>
```
onClickAddRow above needs to be a useCallback.

labkey-bpatel replied:
done

labkey-martyp commented on WNPRC_Purchasing/src/client/components/LineItemsPanelInputs.tsx:
```diff
@@ -67,8 +70,10 @@ export const UnitInput: FC<InputProps> = memo(props => {
 interface NumericInputProps {
     value: number;
-    onChange: (colName, value) => void;
+    onChange?: (colName, value) => void;
```
If you're making this optional you probably want to change the calls below to `onChange?.(...)`

labkey-bpatel replied:
good point, done

labkey-martyp commented on WNPRC_Purchasing/src/client/components/RequestOrderPanel.tsx:
```diff
@@ -60,111 +62,121 @@ export const RequestOrderPanel: FC<Props> = memo(props => {
-            {model.errorMsg && <div className="alert alert-danger">{model.errorMsg}</div>}
-        </Panel>
+        <>
+        {
+            !isAdmin && canUpdate && (
```
Generally put these two on the same line for readability

labkey-bpatel replied:
ok, done

labkey-martyp commented on WNPRC_Purchasing/src/client/components/RequestOrderPanel.tsx:
```diff
@@ -60,111 +62,121 @@ export const RequestOrderPanel: FC<Props> = memo(props => {
+                </Panel>
+            )
+        }
+        {
+            (isAdmin || !canUpdate) && (
```
Same line

labkey-martyp commented on WNPRC_Purchasing/src/client/components/RequestOrderPanelInputs.tsx:
```diff
@@ -86,14 +87,15 @@ export const AccountOtherInput: FC<InputProps> = memo(props => {
 interface VendorInputProps {
-    onChange: (colName, value) => void;
+    onChange?: (colName, value) => void;
     hasError?: boolean;
     model: RequestOrderModel;
-    onModelChange: (model: RequestOrderModel) => void;
+    onModelChange?: (model: RequestOrderModel) => void;
```
Same thing here, if making onChange and onModelChange optional you might want to optionally call them below so you don't get NPE

labkey-bpatel replied:
done

labkey-martyp reviewed (changes requested):
Overall looks great Binal.  Most comments are nitpicky or just considerations.

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/table/WNPRC_PurchasingCustomizer.java:
```diff
@@ -25,18 +26,33 @@ public void customize(TableInfo tableInfo)
 
+    private void addRequestUpdateLinks(AbstractTableInfo ti)
+    {
+        String returnUrl = ti.getUserSchema().getContainer().getPath() + "/WNPRC_Purchasing-purchaseReceiver.view?";
```
Can we use ActionURLs instead of String concatenation here and perhaps for the DetailsURLs too? It looks like the returnUrl is missing the context path as-is.

labkey-bpatel replied:
Good point, now changed!

labkey-jeckels reviewed (changes requested):
One suggestion for the non-JS/TS/TSX code.

labkey-bpatel committed:
Code review changes:
- Use ActionURL to construct detail and return urls
- Instead of setting state for permission variables, use as const vars
- Use async() during initialization
- Extract total cal code into useMemo()
- Cleanup

labkey-bpatel committed:
Fix error handling on line items.

labkey-sweta committed:
Automation test for Receiver's Page

labkey-bpatel reviewed (commented):
Automation looks great, @labkey-sweta
